### PR TITLE
[Feature] Support monocular 3D detection on nuScenes

### PR DESCRIPTION
**Pull request by Tai-Wang**
This PR mainly supports related dataset converter and reading/evaluation API.

### CODE REVIEW

File: mmdet3d/core/bbox/transforms.py [hunk 1]
```diff
@@ -46,13 +46,15 @@ def bbox3d2roi(bbox_list):
     return rois
 
 
-def bbox3d2result(bboxes, scores, labels):
+def bbox3d2result(bboxes, scores, labels, attrs=None):
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/transforms.py:49]: Suggest using `attributes` rather than `attrs`

File: mmdet3d/datasets/__init__.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 from .kitti_dataset import KittiDataset
 from .lyft_dataset import LyftDataset
 from .nuscenes_dataset import NuScenesDataset
+from .nuscenes_dataset_mono import NuScenesDatasetMono
```
Comment (ZwwWayne) [inline mmdet3d/datasets/__init__.py:7]: nuscenes_mono_dataset

File: mmdet3d/core/post_processing/box3d_nms.py [hunk 3]
```diff
@@ -27,10 +29,15 @@ def box3d_multiclass_nms(mlvl_bboxes,
         cfg (dict): Configuration dict of NMS.
         mlvl_dir_scores (torch.Tensor, optional): Multi-level scores
             of direction classifier. Defaults to None.
+        mlvl_attr_scores (torch.Tensor, optional): Multi-level scores
+            of attribute classifier. Defaults to None.
+        mlvl_bboxes2d (torch.Tensor, optional): Multi-level 2D bounding \
```
Comment (ZwwWayne) [inline mmdet3d/core/post_processing/box3d_nms.py:34]: `\` is unnecessary in arguments.

File: mmdet3d/datasets/nuscenes_mono_dataset.py [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+import tempfile
+import torch
+from nuscenes.utils.data_classes import Box as NuScenesBox
+from os import path as osp
+
+from mmdet3d.core import bbox3d2result, box3d_multiclass_nms, xywhr2xyxyr
+from mmdet.datasets import DATASETS
+from ..core.bbox import CameraInstance3DBoxes, get_box_type
+
+try:
```
Comment (ZwwWayne) [inline mmdet3d/datasets/nuscenes_mono_dataset.py:14]: Move coco to where it is imported, see https://github.com/open-mmlab/mmdetection/pull/4880

File: mmdet3d/datasets/nuscenes_mono_dataset.py [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+        if self.eval_version is not None:
+            from nuscenes.eval.detection.config import config_factory
+            self.eval_detection_configs = config_factory(self.eval_version)
+        if self.modality is None:
+            self.modality = dict(
+                use_camera=True,
+                use_lidar=False,
+                use_radar=False,
+                use_map=False,
+                use_external=False,
```
Comment (ZwwWayne) [inline mmdet3d/datasets/nuscenes_mono_dataset.py:101]: clean unnecessary comma

File: mmdet3d/datasets/nuscenes_mono_dataset.py [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+        keep it. Otherwise, we will use the default attribute.
+
+        Args:
+            attr_idx (int): Attribute index.
+            label_name (str): Predicted category name.
+
+        Returns:
+            str: Predicted attribute name.
+        """
+        AttrMapping_rev2 = [
```
Comment (ZwwWayne) [inline mmdet3d/datasets/nuscenes_mono_dataset.py:250]: Why need `rev2`?

File: mmdet3d/datasets/nuscenes_mono_dataset.py [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+                nms_thr=0.05,
+                score_thr=0.01,
+                min_bbox_size=0,
+                max_per_frame=500)
+            from mmcv import Config
+            nms_cfg = Config(nms_cfg)
+            cam_boxes3d_for_nms = xywhr2xyxyr(cam_boxes3d.bev)
+            boxes3d = cam_boxes3d.tensor
+            # generate attr scores from attr labels
+            attrs = torch.LongTensor([attr for attr in attrs_per_frame]).cuda()
```
Comment (ZwwWayne) [inline mmdet3d/datasets/nuscenes_mono_dataset.py:341]: cuda() here is a little bit ugly and dangerous.

File: mmdet3d/datasets/nuscenes_mono_dataset.py (hunk 6 above)
Comment (Tai-Wang) [reply mmdet3d/datasets/nuscenes_mono_dataset.py:250]: `AttrMapping_rev` has been used in the NuScenesDataset. I just use `rev2` to distinguish it with the one used in that class.

File: mmdet3d/datasets/pipelines/transforms_3d.py [hunk 8]
```diff
@@ -116,6 +116,71 @@ def __repr__(self):
         return repr_str
 
 
+@PIPELINES.register_module()
+class RandomFlipMono3D(RandomFlip3D):
```
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/transforms_3d.py:120]: Can we simply enhance the original RandomFlip3D module?

File: mmdet3d/datasets/nuscenes_mono_dataset.py (hunk 6 above)
Comment (ZwwWayne) [reply mmdet3d/datasets/nuscenes_mono_dataset.py:250]: OK, then we may add a TODO in this place to simplify the name.

File: mmdet3d/core/bbox/box_np_ops.py [hunk 9]
```diff
@@ -335,21 +335,40 @@ def rotation_points_single_angle(points, angle, axis=0):
...
         proj_mat (np.ndarray): Transformation matrix between coordinates.
+        with_depth (bool): Whether to keep depth in the output.
 
     Returns:
         np.ndarray: Points in image coordinates with shape [N, 2].
     """
     points_shape = list(points_3d.shape)
     points_shape[-1] = 1
+
+    assert len(proj_mat.shape) == 2, f'The dimension of the projection'\
```
[the remark this answers is not in the document]
Comment (Tai-Wang) [reply mmdet3d/core/bbox/box_np_ops.py]: Also update the corresponding part of the tensor version functions (in the `mmdet3d/core/bbox/structures/utils.py`).

File: mmdet3d/datasets/pipelines/transforms_3d.py (hunk 8 above)
Comment (Tai-Wang) [reply mmdet3d/datasets/pipelines/transforms_3d.py]: Enhanced and merged.

File: mmdet3d/datasets/pipelines/loading.py [hunk 10]
```diff
@@ -65,6 +65,33 @@ def __repr__(self):
...
+class LoadImageFromFileMono3D(LoadImageFromFile):
+    """Load an image from file in monocular 3D object detection. Compared to 2D
+    detection, additional camera parameters need to be loaded.
+
+    Args:
+        kwargs (dict): Arguments are the same as those in \
+            :class:`LoadImageFromFile`.
+    """
+
+    def __init__(self, **kwargs):
```
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/loading.py:78]: lines 78-79 can be ommited.

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**